clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -320,6 +322,25 @@ static bool isSYCLKernelBodyFunction(FunctionDecl *FD) {
   return FD->getOverloadedOperator() == OO_Call;
 }
 
+static bool isSYCLUndefinedAllowed(const FunctionDecl *Callee,
+                                   const SourceManager &SrcMgr) {
+  if (!Callee)
+    return false;
+
+  const Type *Ty = nullptr;
+  // libstdc++-11 introduced undefined function "void __failed_assertion()"
+  // which may lead to SemaSYCL check failure. However, this undefined function
+  // is used to trigger some compilation error when check fails in compilation
+  // time and will be ignored when the check succeeds. We enable this function
+  // to support some important std functions in SYCL device.
+  if (Callee->getName() == GlibcxxFailedAssertion &&
+      (Callee->getNumParams() == 0) &&
+      (Ty = Callee->getReturnType().getTypePtr()) && (Ty->isVoidType()))
+    return SrcMgr.isInSystemHeader(Callee->getLocation());
+
+  return false;
```

Comment:
```suggestion
  return Callee->getName() == GlibcxxFailedAssertion &&
      Callee->getNumParams() == 0 &&
      Callee->getReturnType()->isVoidType() &&
      SrcMgr.isInSystemHeader(Callee->getLocation());
```
Simplifies the logic a little bit.